README.md
## Zoneminder Docker (Deprecated)
(Current version: 1.36)

This container will no longer be maintained or updated.  The current ZM Event Server is being phased out at the end of 2021 and this container will not be updated when a new ZM Event Server is released.

### About
This is an easy to run dockerized image of [ZoneMinder](https://github.com/ZoneMinder/zoneminder) along with the the [ZM Event Notification Server](https://github.com/pliablepixels/zmeventnotification) and its machine learning subsystem.

The configuration settings that are needed for this implementation of Zoneminder are pre-applied and do not need to be changed on the first run of Zoneminder.

This version will now upgrade Zoneminder from previous versions.

You can donate [here](https://www.paypal.com/us/cgi-bin/webscr?cmd=_s-xclick&amp;hosted_button_id=EJGPC7B5CS66E).

### Support
There is minimal support for this docker container at this time.

Go to the Zoneminder Forum [here](https://forums.zoneminder.com/) for support.

### Installation
Install the docker container by going to a command line and enter the command:

```bash
docker pull dlandon/zoneminder.machine.learning
```

This will pull the zoneminder docker image. Once it is installed you are ready to run the docker container.

Before you run the image, feel free to read the configuration section below to customize various settings.

To run Zoneminder:

```bash
docker run -d --name="Zoneminder" \
--net="bridge" \
--privileged="false" \
--shm-size="8G" \
-p 8443:443/tcp \
-p 9000:9000/tcp \
-e TZ="America/New_York" \
-e PUID="99" \
-e PGID="100" \
-e MULTI_PORT_START="0" \
-e MULTI_PORT_END="0" \
-v "/mnt/Zoneminder":"/config":rw \
-v "/mnt/Zoneminder/data":"/var/cache/zoneminder":rw \
dlandon/zoneminder.machine.learning
```

For http:// access use: `-p 8080:80/tcp`

### Shared Memory
Set shared memory to half of your installed memory.

### Subsequent runs

You can start/stop/restart the container anytime. You don't need to run the command above every time. If you have already created the container once (by the `docker run` command above), you can simply do a `docker stop Zoneminder` to stop it and a `docker start Zoneminder` to start it any time (or do a `docker restart Zoneminder`).

#### Customization

- Set `MULTI_PORT_START` and `MULTI_PORT_END` to define a port range for ES multi-port operation.
- The commands above use a host path of `/mnt/Zoneminder` to map the container config and cache directories. This is going to be persistent directory that will retain data across container/image stop/restart/deletes. ZM mysql/other config data/event files/etc are kept here. You can change this to any directory in your host path that you want to.

#### User Script

You can enable a custom user script that will run every time the container is started.

Put your script in the /mnt/Zoneminder/ folder and name it userscript.sh. The script will be executed each time the container is started before Zoneminder is started. Be sure to `chmod +x userscript.sh` so the script is executable. Set `ADVANCED_SCRIPT="1"` environment variable to enable your script.

#### Adding Nvidia GPU support to the Zoneminder.

You will have to install support for your graphics card. If you are using Unraid, install the Nvidia plugin and follow these [instructions](https://forums.unraid.net/topic/77813-plugin-linuxserverio-unraid-nvidia/?tab=comments#comment-719665). On other systems install the Nvidia docker, see [here](https://medium.com/@adityathiruvengadam/cuda-docker-%EF%B8%8F-for-deep-learning-cab7c2be67f9).

After you confirm the graphics card is seen by the Zoneminder docker container, you can then compile opencv with GPU support. Be sure your container can see the graphics card. Read the `opencv.sh` script for instructions on how to download the packages needed for compiling opencv. You will need to get a developer account for some of the packages because of licensing. Get into the docker command line by `docker exec -it Zoneminder /bin/bash` and do this once you have the packages:

```bash
cd /config/opencv
./opencv.sh
```

This will compile the opencv with GPU support. It takes a LONG time. You should then have GPU support.

You will have to install the CuDNN runtime yourself based on your particular setup.

#### Post install configuration and caveats

- After successful installation, please refer to the [ZoneMinder](https://zoneminder.readthedocs.io/en/stable/), [Event Server and Machine Learning](https://zmeventnotification.readthedocs.io/en/latest/index.html) configuration guides from the authors of these components to set it up to your needs. Specifically, if you are using the Event Server and the machine learning hooks, you will need to customize `/etc/zm/zmeventnotification.ini` and `/etc/zm/objectconfig.ini`.

- Note that by default, this docker build runs ZM on port 443 inside the docker container and maps it to port 8443 for the outside world. Therefore, if you are configuring `/etc/zm/objectconfig.ini` or `/etc/zm/zmeventnotification.ini` remember to use `https://localhost:443/<etc>` as the base URL.

- Push notifications with images will not work unless you replace the self-signed certificates that are auto-generated. Feel free to use the excellent and free [LetsEncrypt](https://letsencrypt.org) service if you'd like.

#### Usage

To access the Zoneminder GUI, browse to: `https://<your host ip>:8443/zm` or `http://<your host ip>:8080/zm` if `-p 8080:80/tcp` is specified.

The zmNinja Event Notification Server is accessed at port `9000`. Security with a self signed certificate is enabled. You may have to install the certificate on iOS devices for the event notification to work properly.

#### Troubleshooting when the container fails

If you have a situation where the container fails to start, you can set `NO_START_ZM="1"` as an environment variable - this will spin up the container but will not automatically start the MySql and Zoneminder processes. This way, you can get into a command line in the container (`docker exec -it Zoneminder /bin/bash`) and troubleshoot your issue by using the following commands to start MySql and Zoneminder and fix any errors/problems with them starting.

```bash
service mysql start
service zoneminder start
```

CHANGELOG.md
## Change Log

### 2023-05-20
- Update Ubuntu to 22.04
- Update zmNinja Event Notification Server to version 6.1.28.
- Update Opencv to 4.7.0

### 2021.10.29
- Fix update issue with mariadb.

### 2021.10.08
- Delay start of Zoneminder on first run.
- Update Opencv to 4.5.3.

### 2021.06.30
- Apply a fix for database failure on first run.

### 2021.05.30
- Fix for Zoneminder 1.36.3.
- Update to opencv 4.5.2.

### 2021-05-23
- Update to Zoneminder 1.36.

### 2021-05-20
- Update zmNinja Event Notification Server to version 6.1.23.

### 2021-04-18
- Update zmNinja Event Notification Server to version 6.1.22.

### 2021-04-08
- Update zmNinja Event Notification Server to version 6.1.19.

### 2021-04-04
- Update zmNinja Event Notification Server to version 6.1.18.

### 2021-03-27
- Update zmNinja Event Notification Server to version 6.1.16.

### 2021-03-18
- Update zmNinja Event Notification Server to version 6.1.15.

### 2021-02-16
- Update zmNinja Event Notification Server to version 6.1.12.

### 2021-01-30
- Update zmNinja Event Notification Server to version 6.1.11.

### 2021-01-26
- Update zmNinja Event Notification Server to version 6.1.10.

### 2021-01-24
- Preliminary release of machine learning Docker.

### 2021-01-19
- Prepare for Zoneminder ES/ML Docker container with ES/ML preloaded.

### 2021-01-07
- Update opencv.sh to Ubuntu 20.04.
- Remove /config/mysql/ib_logfile* files to insure that mysql starts.
- Increase shm-size to 8G.

### 2021-01-03
- Fix syslog-ng configuration file version.
- Update zmNinja Event Notification Server to version 6.1.5.

### 2021-01-02
- Update Docker baseimage to Focal Alpha (Ubuntu 20.04).
- Remove SHMEM environment variable and set --shm-size instead.
- Turn off privileged mode.
- Update zmNinja Event Notification Server to version 6.1.0.

### 2020-11-18
- Add NO_START_ZM environment variable to keep MySql and Zoneminder from starting so the docker stays running and a user can troubleshoot.

### 2020-10-24
- Update zmNinja Event Notification Server to version 6.0.5.
- Update opencv to 4.3.
- Improve onepcv compile time.

### 2020-10-17
- Update zmNinja Event Notification Server to version 6.0.4.

### 2020-08-12
- Fix ES Yolo model download destination file names.
- Create unknown_faces folder.

### 2020-08-11
- Modify ES model downloads to the new paths and add Yolo V4.
- Changed Yolo model download environment variables.

### 2020-08-08
- Update zmNinja Event Notification Server to version 5.15.6.

###2020-07-19
- Update baseimage to bionic-1.0.0.

### 2020-07-03
- Add unknown_faces folder and symlink.

### 2020-06-22
- Update zmNinja Event Notification Server to version 5.15.5.
- Add pushover api.

### 2020-04-03
- Update zmNinja Event Notification Server to version 5.11.3.

### 2020-03-08
- Add multi-port ES capablity to adjust apache2 for multi-port operation.

### 2020-02-29
- Fix docker failure when it can't update.

### 2020-02-23
- Fix reboot apt update of cuda failing again.
- Set up unattended re-compile of opencv when docker updates.
- More opencv script changes.

### 2020-02-22
- Move all opencv stuff to /config/opencv/ folder.
- Fix reboot apt update of cuda failing.

### 2020-02-21
- Fix uninstall and install of hooks when updating ES.

### 2020-02-20
- Update zmNinja Event Notification Server to version 5.7.7.

### 2020-02-17
- Change detection of INSTALL_HOOK already being installed,

### 2020-02-16
- Update zmNinja Event Notification Server to version 5.7.4 and add opencv.sh compile script.
- Modify ES update to install a tar to update ES so a new docker does not have to be built whrn ES is updated.

### 2020-02-05
- Adjust /var/cache/zoneminder ownership to www-data:www-data.

### 2020-01-17
- Update Zoneminder to 1.34.

### 2020-01-10
- Update zmNinja Event Notification Server to version 5.4.

### 2019-12-01
- Fix ES file references to newer versions of zm_detect_wrapper.sh amnd zm_detect.py

### 2019-11-23
- Update zmNinja Event Notification Server to version 4.6.

### 2019-11-09
- Update zmNinja Event Notification Server to version 4.5.

### 2019-10-26
- Remove cambozola legacy browser support.

### 2019-10-05
- Update zmNinja Event Notification Server to version 4.4.

### 2019-09-21
- Fix /hook folder permission check.
- Remove ZM_PATH_ZMS update from defaults.

### 2019-09-19
- Update zmes_hook_helpers.

### 2019-09-17
- Additional changes for zmevent server installation.

### 2019-09-15
- Update README to make it more generic.
- Add INSTALL_YOLO and INSTALL_TINY_YOLO environment variables to download the model files for zmeventserver when hook processing is enabled.
- Update zmNinja Event Notification Server to version 4.2.

### 2019-09-05
- Fix update script.

### 2019-08-05
- Added pyzmutils and requests modules for hook processing.

### 2019-08-03
- Use pip3 for setup.py install of hook processing.

### 2019-07-30
- Update zmNinja Event Notification Server to version 4.1.

### 2019-07-19
- Update zmNinja Event Notification Server to version 3.9.

### 2019-06-09
- Update zmNinja Event Notification Server to version 3.8.

### 2019-05-12
- Update zmNinja Event Notification Server to version 3.7.

### 2019-04-27
- Add Net::MQTT::Simple::Auth perl library.

### 2019-04-26
- Add "INSTALL_FACE" environment variable to load face recognition package.

### 2019-04-25
- Add zmes_hook_helpers to docker image and change the hook installation.

### 2019-04-24
- Update zmNinja Event Notification Server to version 3.6.
- Add "INSTALL_HOOK" environment variable to load packages and run 'setup.py' for hook processing.

### 2019-04-21
- Update zmNinja Event Notification Server to version 3.5.

### 2019-04-16
- Update zmNinja Event Notification Server to version 3.4.

### 2019-04-05
- Fix: Minor adjustments to zmeventnotification.

### 2019-03-30
- Fix: Control file copy was copying the wrong file extension.

### 2019-03-27
- Add: Install vlc packages.

### 2019-03-25
- Update zmNinja Event Notification Server to version 3.3.

### 2019-03-04

init/40_firstrun.sh
#!/bin/bash
#
# 40_firstrun.sh
#
#
# Github URL for opencv zip file download.
# Current default is to pull the version 4.3.0 release.
#
# Search for config files, if they don't exist, create the default ones
if [ ! -d /config/conf ]; then
	echo "Creating conf folder"
	mkdir /config/conf
else
	echo "Using existing conf folder"
fi

# Handle the zm.conf files
echo "Copying zm.conf to config folder"
mv /root/zm.conf /config/conf/zm.default
cp /etc/zm/conf.d/README /config/conf/README

# Copy custom 99-mysql.conf to /etc/zm/conf.d/
if [ -f /config/conf/99-mysql.conf ]; then
	echo "Copy custom 99-mysql.conf to /etc/zm/conf.d/"
	cp /config/conf/99-mysql.conf /etc/zm/conf.d/99-mysql.conf
fi

# Handle the zmeventnotification.ini file
if [ -f /root/zmeventnotification/zmeventnotification.ini ]; then
	echo "Moving zmeventnotification.ini"
	cp /root/zmeventnotification/zmeventnotification.ini /config/zmeventnotification.ini.default
	if [ ! -f /config/zmeventnotification.ini ]; then
		mv /root/zmeventnotification/zmeventnotification.ini /config/zmeventnotification.ini
	else
		rm -rf /root/zmeventnotification/zmeventnotification.ini
	fi
else
	echo "File zmeventnotification.ini already moved"
fi

# Handle the secrets.ini file
if [ -f /root/zmeventnotification/secrets.ini ]; then
	echo "Moving secrets.ini"
	cp /root/zmeventnotification/secrets.ini /config/secrets.ini.default
	if [ ! -f /config/secrets.ini ]; then
		mv /root/zmeventnotification/secrets.ini /config/secrets.ini
	else
		rm -rf /root/zmeventnotification/secrets.ini
	fi
else
	echo "File secrets.ini already moved"
fi

# Create opencv folder if it doesn't exist
if [ ! -d /config/opencv ]; then
	echo "Creating opencv folder in config folder"
	mkdir /config/opencv
fi

# Handle the opencv.sh file
if [ -f /root/zmeventnotification/opencv.sh ]; then
	echo "Moving opencv.sh"
	cp /root/zmeventnotification/opencv.sh /config/opencv/opencv.sh.default
	if [ ! -f /config/opencv/opencv.sh ]; then
		mv /root/zmeventnotification/opencv.sh /config/opencv/opencv.sh
	else
		rm -rf /root/zmeventnotification/opencv.sh
	fi
else
	echo "File opencv.sh already moved"
fi

# Handle the debug_opencv.sh file
if [ -f /root/zmeventnotification/debug_opencv.sh ]; then
	echo "Moving debug_opencv.sh"
	mv /root/zmeventnotification/debug_opencv.sh /config/opencv/debug_opencv.sh
else
	echo "File debug_opencv.sh already moved"
fi

if [ ! -f /config/opencv/opencv_ok ]; then
	echo "no" > /config/opencv/opencv_ok
fi

# Handle the zmeventnotification.pl
if [ -f /root/zmeventnotification/zmeventnotification.pl ]; then
	echo "Moving the event notification server"
	mv /root/zmeventnotification/zmeventnotification.pl /usr/bin
	chmod +x /usr/bin/zmeventnotification.pl 2>/dev/null
else
	echo "Event notification server already moved"
fi

# Handle the pushapi_pushover.py
if [ -f /root/zmeventnotification/pushapi_pushover.py ]; then
	echo "Moving the pushover api"
	mkdir -p /var/lib/zmeventnotification/bin/
	mv /root/zmeventnotification/pushapi_pushover.py /var/lib/zmeventnotification/bin/
	chmod +x /var/lib/zmeventnotification/bin/pushapi_pushover.py 2>/dev/null
else
	echo "Pushover api already moved"
fi

# Handle the es_rules.json
if [ -f /root/zmeventnotification/es_rules.json ]; then
	echo "Moving es_rules.json"
	cp /root/zmeventnotification/es_rules.json /config/es_rules.json.default
	if [ ! -f /config/es_rules.json ]; then
		mv /root/zmeventnotification/es_rules.json /config/es_rules.json
	else
		rm -rf /etc/zm/es_rules.json
	fi
fi

# Move ssmtp configuration if it doesn't exist
if [ ! -d /config/ssmtp ]; then
	echo "Moving ssmtp to config folder"
	cp -p -R /etc/ssmtp/ /config/
else
	echo "Using existing ssmtp folder"
fi

# Move mysql database if it doesn't exit
if [ ! -d /config/mysql/mysql ]; then
	echo "Moving mysql to config folder"
	rm -rf /config/mysql
	cp -p -R /var/lib/mysql /config/
else
	echo "Using existing mysql database folder"
fi

# files and directories no longer exposed at config.
rm -rf /config/perl5/
rm -rf /config/zmeventnotification/
rm -rf /config/zmeventnotification.pl
rm -rf /config/skins
rm -rf /config/zm.conf

# Create Control folder if it doesn't exist and copy files into image
if [ ! -d /config/control ]; then
	echo "Creating control folder in config folder"
	mkdir /config/control
else
	echo "Copy /config/control/ scripts to /usr/share/perl5/ZoneMinder/Control/"
	cp /config/control/*.pm /usr/share/perl5/ZoneMinder/Control/ 2>/dev/null
	chown root:root /usr/share/perl5/ZoneMinder/Control/* 2>/dev/null
	chmod 644 /usr/share/perl5/ZoneMinder/Control/* 2>/dev/null
fi

# Copy conf files if there are any
if [ -d /config/conf ]; then
	echo "Copy /config/conf/ scripts to /etc/zm/conf.d/"
	cp /config/conf/*.conf /etc/zm/conf.d/ 2>/dev/null
	chown root:root /etc/zm/conf.d* 2>/dev/null
	chmod 640 /etc/conf.d/* 2>/dev/null
fi

echo "Creating symbolink links"
# security certificate keys
rm -rf /etc/apache2/ssl/zoneminder.crt
ln -sf /config/keys/cert.crt /etc/apache2/ssl/zoneminder.crt
rm -rf /etc/apache2/ssl/zoneminder.key
ln -sf /config/keys/cert.key /etc/apache2/ssl/zoneminder.key
mkdir -p /var/lib/zmeventnotification/push
mkdir -p /config/push
rm -rf /var/lib/zmeventnotification/push/tokens.txt
ln -sf /config/push/tokens.txt /var/lib/zmeventnotification/push/tokens.txt
ln -sf /config/es_rules.json /etc/zm/es_rules.json

# ssmtp
rm -r /etc/ssmtp 
ln -s /config/ssmtp /etc/ssmtp

# mysql
rm -r /var/lib/mysql
ln -s /config/mysql /var/lib/mysql

# Set ownership for unRAID
PUID=${PUID:-99}
PGID=${PGID:-100}
usermod -o -u $PUID nobody

# Check if the group with GUID passed as environment variable exists and create it if not.
if ! getent group "$PGID" >/dev/null; then
  groupadd -g "$PGID" env-provided-group
  echo "Group with id: $PGID did not already exist, so we created it."
fi

usermod -g $PGID nobody
usermod -d /config nobody

# Set ownership for mail
usermod -a -G mail www-data

# Change some ownership and permissions
chown -R mysql:mysql /config/mysql
chown -R mysql:mysql /var/lib/mysql
chown -R $PUID:$PGID /config/conf
chmod 777 /config/conf
chmod 666 /config/conf/*
chown -R $PUID:$PGID /config/control
chmod 777 /config/control
chmod 666 -R /config/control/
chown -R $PUID:$PGID /config/ssmtp
chmod -R 777 /config/ssmtp
chown -R $PUID:$PGID /config/zmeventnotification.*
chmod 666 /config/zmeventnotification.*
chown -R $PUID:$PGID /config/secrets.ini
chmod 666 /config/secrets.ini
chown -R $PUID:$PGID /config/opencv
chmod 777 /config/opencv
chmod 666 /config/opencv/*
chown -R $PUID:$PGID /config/keys
chmod 777 /config/keys
chmod 666 /config/keys/*
chown -R www-data:www-data /config/push/
chown -R www-data:www-data /var/lib/zmeventnotification/
chmod +x /config/opencv/opencv.sh
chmod +x /config/opencv/debug_opencv.sh
chmod +x /config/opencv/opencv.sh.default

# Create events folder
if [ ! -d /var/cache/zoneminder/events ]; then
	echo "Create events folder"
	mkdir /var/cache/zoneminder/events
	chown -R www-data:www-data /var/cache/zoneminder/events
	chmod -R 777 /var/cache/zoneminder/events
else
	echo "Using existing data directory for events"

	# Check the ownership on the /var/cache/zoneminder/events directory
	if [ `stat -c '%U:%G' /var/cache/zoneminder/events` != 'www-data:www-data' ]; then
		echo "Correcting /var/cache/zoneminder/events ownership..."
		chown -R www-data:www-data /var/cache/zoneminder/events
	fi

	# Check the permissions on the /var/cache/zoneminder/events directory
	if [ `stat -c '%a' /var/cache/zoneminder/events` != '777' ]; then
		echo "Correcting /var/cache/zoneminder/events permissions..."
		chmod -R 777 /var/cache/zoneminder/events
	fi
fi

# Create images folder
if [ ! -d /var/cache/zoneminder/images ]; then
	echo "Create images folder"
	mkdir /var/cache/zoneminder/images
	chown -R www-data:www-data /var/cache/zoneminder/images
	chmod -R 777 /var/cache/zoneminder/images
else
	echo "Using existing data directory for images"

	# Check the ownership on the /var/cache/zoneminder/images directory
	if [ `stat -c '%U:%G' /var/cache/zoneminder/images` != 'www-data:www-data' ]; then
		echo "Correcting /var/cache/zoneminder/images ownership..."
		chown -R www-data:www-data /var/cache/zoneminder/images
	fi

	# Check the permissions on the /var/cache/zoneminder/images directory
	if [ `stat -c '%a' /var/cache/zoneminder/images` != '777' ]; then
		echo "Correcting /var/cache/zoneminder/images permissions..."
		chmod -R 777 /var/cache/zoneminder/images
	fi
fi

# Create temp folder
if [ ! -d /var/cache/zoneminder/temp ]; then
	echo "Create temp folder"
	mkdir /var/cache/zoneminder/temp
	chown -R www-data:www-data /var/cache/zoneminder/temp
	chmod -R 777 /var/cache/zoneminder/temp
else
	echo "Using existing data directory for temp"

	# Check the ownership on the /var/cache/zoneminder/temp directory
	if [ `stat -c '%U:%G' /var/cache/zoneminder/temp` != 'www-data:www-data' ]; then
		echo "Correcting /var/cache/zoneminder/temp ownership..."
		chown -R www-data:www-data /var/cache/zoneminder/temp
	fi

	# Check the permissions on the /var/cache/zoneminder/temp directory
	if [ `stat -c '%a' /var/cache/zoneminder/temp` != '777' ]; then
		echo "Correcting /var/cache/zoneminder/temp permissions..."
		chmod -R 777 /var/cache/zoneminder/temp
	fi
fi

# Create cache folder
if [ ! -d /var/cache/zoneminder/cache ]; then
	echo "Create cache folder"
	mkdir /var/cache/zoneminder/cache
	chown -R www-data:www-data /var/cache/zoneminder/cache
	chmod -R 777 /var/cache/zoneminder/cache
else
	echo "Using existing data directory for cache"

	# Check the ownership on the /var/cache/zoneminder/cache directory
	if [ `stat -c '%U:%G' /var/cache/zoneminder/cache` != 'www-data:www-data' ]; then
		echo "Correcting /var/cache/zoneminder/cache ownership..."
		chown -R www-data:www-data /var/cache/zoneminder/cache
	fi

	# Check the permissions on the /var/cache/zoneminder/cache directory
	if [ `stat -c '%a' /var/cache/zoneminder/cache` != '777' ]; then
		echo "Correcting /var/cache/zoneminder/cache permissions..."
		chmod -R 777 /var/cache/zoneminder/cache
	fi
fi

# set user crontab entries
crontab -r -u root
if [ -f /config/cron ]; then
	crontab -l -u root | cat - /config/cron | crontab -u root -
fi

# Symbolink for /config/zmeventnotification.ini
ln -sf /config/zmeventnotification.ini /etc/zm/zmeventnotification.ini
chown www-data:www-data /etc/zm/zmeventnotification.ini

# Symbolink for /config/secrets.ini
ln -sf /config/secrets.ini /etc/zm/

# Set multi-ports in apache2 for ES.
# Start with default configuration.
cp /etc/apache2/ports.conf.default /etc/apache2/ports.conf
cp /etc/apache2/sites-enabled/default-ssl.conf.default /etc/apache2/sites-enabled/default-ssl.conf

if [ $((MULTI_PORT_START)) -gt 0 ] && [ $((MULTI_PORT_END)) -gt $((MULTI_PORT_START)) ]; then

	echo "Setting ES multi-port range from ${MULTI_PORT_START} to ${MULTI_PORT_END}."

	ORIG_VHOST="_default_:443"

	NEW_VHOST=${ORIG_VHOST}
	PORT=${MULTI_PORT_START}
	while [[ ${PORT} -le ${MULTI_PORT_END} ]]; do
	    egrep -sq "Listen ${PORT}" /etc/apache2/ports.conf || echo "Listen ${PORT}" >> /etc/apache2/ports.conf
	    NEW_VHOST="${NEW_VHOST} _default_:${PORT}"
	    PORT=$(($PORT + 1))
	done

	perl -pi -e "s/${ORIG_VHOST}/${NEW_VHOST}/ if (/<VirtualHost/);" /etc/apache2/sites-enabled/default-ssl.conf
else
	if [ $((MULTI_PORT_START)) -ne 0 ];then
		echo "Multi-port error start ${MULTI_PORT_START}, end ${MULTI_PORT_END}."
	fi
fi

# Move the yolo models to /var/lib/zmeventnotification
if [ -d /root/models ]; then
	mv /root/models/ /var/lib/zmeventnotification/
	chown -R www-data:www-data /var/lib/zmeventnotification/models
fi

# Create hook folder
if [ ! -d /config/hook ]; then
	echo "Creating /config/hook folder"
	mkdir /config/hook
fi

# Handle the objectconfig.ini file
if [ -f /root/zmeventnotification/objectconfig.ini ]; then
	echo "Moving objectconfig.ini"
	cp /root/zmeventnotification/objectconfig.ini /config/hook/objectconfig.ini.default
	if [ ! -f /config/hook/objectconfig.ini ]; then
		mv /root/zmeventnotification/objectconfig.ini /config/hook/objectconfig.ini
	else
		rm -rf /root/zmeventnotification/objectconfig.ini
	fi
else
	echo "File objectconfig.ini already moved"
fi

# Handle the config_upgrade script
if [ -f /root/zmeventnotification/config_upgrade.py ]; then
	echo "Moving config_upgrade.py"
	mv /root/zmeventnotification/config_upgrade.py /config/hook/config_upgrade.py
	rm -rf /config/hook/config_upgrade.sh
	chmod +x /config/hook/config_upgrade.*
else
	echo "File config_upgrade.py already moved"
fi

# Handle the zm_event_start.sh file
if [ -f /root/zmeventnotification/zm_event_start.sh ]; then
	echo "Moving zm_event_start.sh"
	mv /root/zmeventnotification/zm_event_start.sh /config/hook/zm_event_start.sh
else
	echo "File zm_event_start.sh already moved"
fi

# Handle the zm_event_end.sh file
if [ -f /root/zmeventnotification/zm_event_end.sh ]; then
	echo "Moving zm_event_end.sh"
	mv /root/zmeventnotification/zm_event_end.sh /config/hook/zm_event_end.sh
else
	echo "File zm_event_end.sh already moved"
fi

# Handle the zm_detect.py file
if [ -f /root/zmeventnotification/zm_detect.py ]; then
	echo "Moving zm_detect.py"
	mv /root/zmeventnotification/zm_detect.py /config/hook/zm_detect.py
else
	echo "File zm_detect.py already moved"
fi

# Handle the zm_detect_old.py file
if [ -f /root/zmeventnotification/zm_detect_old.py ]; then
	echo "Moving zm_detect_old.py"
	mv /root/zmeventnotification/zm_detect_old.py /config/hook/zm_detect_old.py
else
	echo "File zm_detect_old.py already moved"
fi

# Handle the zm_train_faces.py file
if [ -f /root/zmeventnotification/zm_train_faces.py ]; then
	echo "Moving zm_train_faces.py"
	mv /root/zmeventnotification/zm_train_faces.py /config/hook/zm_train_faces.py
else
	echo "File zm_train_faces.py already moved"
fi

# Handle the train_faces.py file
if [ -f /root/zmeventnotification/train_faces.py ]; then
	echo "Moving train_faces.py"
	mv /root/zmeventnotification/train_faces.py /config/hook/train_faces.py
else
	echo "File train_faces.py already moved"
fi

# Symbolic link for known_faces in /config
rm -rf /var/lib/zmeventnotification/known_faces
ln -sf /config/hook/known_faces /var/lib/zmeventnotification/known_faces
chown -R www-data:www-data /var/lib/zmeventnotification/known_faces

# Symbolic link for unknown_faces in /config
rm -rf /var/lib/zmeventnotification/unknown_faces
ln -sf /config/hook/unknown_faces /var/lib/zmeventnotification/unknown_faces
chown -R www-data:www-data /var/lib/zmeventnotification/unknown_faces

# Symbolic link for misc in /config
rm -rf /var/lib/zmeventnotification/misc
ln -sf /config/hook/misc /var/lib/zmeventnotification/misc
chown -R www-data:www-data /var/lib/zmeventnotification/misc

# Create misc folder if it doesn't exist
if [ ! -d /config/hook/misc ]; then
	echo "Creating hook/misc folder in config folder"
	mkdir -p /config/hook/misc
fi

# Create coral_edgetpu folder if it doesn't exist
if [ ! -d /config/hook/coral_edgetpu ]; then
	echo "Creating hook/coral_edgetpu folder in config folder"
	mkdir -p /config/hook/coral_edgetpu
fi

# Symbolic link for coral_edgetpu in /config
rm -rf /var/lib/zmeventnotification/models/coral_edgetpu
ln -sf /config/hook/coral_edgetpu/ /var/lib/zmeventnotification/models
chown -R www-data:www-data /var/lib/zmeventnotification/models/coral_edgetpu 2>/dev/null

# Symbolic link for hook files in /config
mkdir -p /var/lib/zmeventnotification/bin
ln -sf /config/hook/zm_detect.py /var/lib/zmeventnotification/bin/zm_detect.py
ln -sf /config/hook/zm_detect_old.py /var/lib/zmeventnotification/bin/zm_detect_old.py
ln -sf /config/hook/zm_train_faces.py /var/lib/zmeventnotification/bin/zm_train_faces.py
ln -sf /config/hook/train_faces.py /var/lib/zmeventnotification/bin/train_faces.py
ln -sf /config/hook/zm_event_start.sh /var/lib/zmeventnotification/bin/zm_event_start.sh
ln -sf /config/hook/zm_event_end.sh /var/lib/zmeventnotification/bin/zm_event_end.sh
chmod +x /var/lib/zmeventnotification/bin/*
ln -sf /config/hook/objectconfig.ini /etc/zm/

# Create known_faces folder if it doesn't exist
if [ ! -d /config/hook/known_faces ]; then
	echo "Creating hook/known_faces folder in config folder"
	mkdir -p /config/hook/known_faces
fi

# Create unknown_faces folder if it doesn't exist
if [ ! -d /config/hook/unknown_faces ]; then
	echo "Creating hook/unknown_faces folder in config folder"
	mkdir -p /config/hook/unknown_faces
fi

# Set hook folder permissions
chown -R $PUID:$PGID /config/hook
chmod -R 777 /config/hook

# Compile opencv
if [ -f /config/opencv/opencv_ok ] && [ `cat /config/opencv/opencv_ok` = 'yes' ]; then
	if [ ! -f /root/setup.py ]; then
		if [ -x /config/opencv/opencv.sh ]; then
			/config/opencv/opencv.sh quiet &>/dev/null
		fi
	fi
fi

mv /root/zmeventnotification/setup.py /root/setup.py 2>/dev/null

# Clean up mysql log files to insure mysql will start
rm -f /config/mysql/ib_logfile* 2>/dev/null

echo "Starting services..."
service apache2 start
if [ "$NO_START_ZM" != "1" ]; then
	# Start mysql
	sleep 3
	service mysql restart

	# Update the database if necessary
	zmupdate.pl -nointeractive
	zmupdate.pl -f

	sleep 15
	service zoneminder start
else
	echo "MySql and Zoneminder not started."
fi
